crates/ruma-events/src/message.rs

```diff
@@ -0,0 +1,114 @@
+//! Types for extensible text message events ([MSC1767]).
+//!
+//! [MSC1767]: https://github.com/matrix-org/matrix-doc/pull/1767
+
+use serde::{Deserialize, Serialize};
+
+mod content_serde;
+
+use content_serde::MessageContentSerDeHelper;
+
+/// Text message content.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Text {
+    /// The mime type of the `body`.
+    #[serde(default = "Text::default_mimetype")]
+    pub mimetype: String,
+
+    /// The text content.
+    pub body: String,
+}
+
+impl Text {
+    /// Creates a new plain text message body.
+    pub fn plain(body: impl Into<String>) -> Self {
+        Self { mimetype: "text/plain".to_owned(), body: body.into() }
+    }
+
+    /// Creates a new HTML-formatted message body.
+    pub fn html(body: impl Into<String>) -> Self {
+        Self { mimetype: "text/html".to_owned(), body: body.into() }
+    }
+
+    /// Creates a new HTML-formatted message body by parsing the markdown in `body`.
+    ///
+    /// Returns `None` if no markdown formatting was found.
+    #[cfg(feature = "markdown")]
+    pub fn markdown(body: impl AsRef<str>) -> Option<Self> {
+        let body = body.as_ref();
+        let mut html_body = String::new();
+
+        pulldown_cmark::html::push_html(&mut html_body, pulldown_cmark::Parser::new(body));
+
+        (html_body != format!("<p>{}</p>\n", body)).then(|| Self::html(html_body))
+    }
+
+    fn default_mimetype() -> String {
+        "text/plain".to_owned()
+    }
+}
+
+/// Text message content.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(try_from = "MessageContentSerDeHelper", into = "MessageContentSerDeHelper")]
+pub struct MessageContent(pub(crate) Vec<Text>);
+
+impl MessageContent {
+    /// Get the plain text representation of this message.
+    pub fn plain(&self) -> Option<&str> {
```

Comment:
I would like to "reverse" the naming here and use `plain` and `html` for the constructors, then `find_plain` and `find_html` for these accessors (since `find` more clearly tells that this does an actual search).